In [1]:
from pathlib import Path, PurePath
import glob
import os
import numpy as np
import csv
import pandas as pd
import time
import matplotlib.pyplot as plt

In [2]:
file_path = PurePath(os.getcwd())
dataset_path = file_path.parent.parent.parent.joinpath('eicu')

In [3]:
lab_file = pd.read_csv(dataset_path.joinpath('lab.csv'), usecols=[
    'patientunitstayid', 'labresultoffset', 'labname', 'labresultrevisedoffset'])

d_lab_file = pd.read_csv(dataset_path.joinpath('lab_delirium.csv'), usecols=[
    'patientunitstayid', 'labresultoffset', 'labname', 'labresultrevisedoffset'])


In [8]:
# lab_file = lab_file[['patientunitstayid', 'labresultoffset', 'labname', 'labresultrevisedoffset']]
# d_lab_file = d_lab_file[['patientunitstayid', 'labresultoffset', 'labname', 'labresultrevisedoffset']]

'labid', 'patientunitstayid', 'labresultoffset', 'labtypeid', 'labname', 'labresult','labresulttext', 'labmeasurenamesystem', 'labmeasurenameinterface', 'labresultrevisedoffset'

In [4]:
unique_ids = lab_file['patientunitstayid'].unique()
d_unique_ids = d_lab_file['patientunitstayid'].unique()
unique_labs = lab_file['labname'].unique()

In [10]:
print(unique_ids[0:10])
print(d_unique_ids[0:10])
print(unique_labs[0:10])

[141168 141178 141179 141194 141196 141197 141203 141208 141227 141229]
[149321 150934 151213 151708 152171 155126 156970 162054 163084 164338]
['fibrinogen' 'PT - INR' 'magnesium' 'PT' 'pH' 'urinary creatinine' 'MCH'
 'BUN' 'urinary sodium' 'HCO3']


In [12]:


# for id in unique_ids:

#     df_id = lab_file[lab_file['patientunitstayid'] == id]

#     for lab in unique_labs:
        
#         df_lab = df_id[df_id['labname'] == lab]
#         # df_lab = df_lab.sort_values(axis=0, by='labresultoffset')
#         # print(df_lab)
#         times = np.sort(df_lab['labresultoffset'].to_numpy())
#         rev_times = np.sort(df_lab['labresultrevisedoffset'].to_numpy())
#         count = len(times)
#         time_diff = times[1:] - times[:-1]
#         time_diff = time_diff[time_diff != 0]
#         rev_time_diff = rev_times[1:] - rev_times[:-1]
#         rev_time_diff = rev_time_diff[rev_time_diff != 0]

        

In [66]:
df_lab = lab_file[lab_file['labname'] == 'fibrinogen']
# print(df_lab[df_lab['patientunitstayid'] == 2074269])
lab_count = df_lab['patientunitstayid'].value_counts()
print(lab_count.median())

1.0


In [18]:
def get_stats(name, arr):
    if arr.empty:
        return [name, np.NaN, np.NaN, np.NaN, np.NaN, np.NaN, np.NaN, np.NaN]
    else:
        q3, q1 = np.percentile(arr, [75, 25])
        IQR = q3 - q1
    # mean = arr.mean()
    # median = arr.median()
    # var = arr.var()
    # minima = arr.min()
    # maxima = arr.max()
    # return mean, var, median, IQR, minima, maxima
    return [name, arr.mean(), arr.var(), np.median(arr), IQR, arr.min(), arr.max(), len(arr)]

In [55]:
lab_count_arr = []
for lab in unique_labs:

    df_lab = lab_file[lab_file['labname'] == lab]
    lab_count = df_lab['patientunitstayid'].value_counts()
    stats = get_stats(lab, lab_count)
    lab_count_arr.append(stats)

count_df = pd.DataFrame(lab_count_arr, columns = ["lab", "mean", "var", "median", "IQR", "min", "max", "n"])

1.3491873741149902
1.3704020977020264
1.3802015781402588
1.3316471576690674
1.3416483402252197
1.283470630645752
1.5045974254608154
1.5130631923675537
1.3143110275268555
1.4126169681549072
1.3567440509796143
1.4446380138397217
1.4272899627685547
1.3690106868743896
1.4017069339752197
1.3119606971740723
1.3828706741333008
1.4640965461730957
1.3645148277282715
1.2930963039398193
1.3722507953643799
1.3711206912994385
1.4041738510131836
1.4383528232574463
1.3251190185546875
1.4813141822814941
1.3147954940795898
1.4885170459747314
1.4788377285003662
1.4409613609313965
1.3342795372009277
1.42588210105896
1.4629340171813965
1.4360263347625732
1.4982011318206787
1.4160194396972656
1.4605154991149902
1.3802988529205322
1.3728954792022705
1.3159880638122559
1.3867683410644531
1.3725473880767822
1.3493456840515137
1.3405966758728027
1.4699959754943848
1.362262487411499
1.4302804470062256
1.3443658351898193
1.426342487335205
1.360633134841919
1.3004498481750488
1.282400131225586
1.3036487102508545


KeyboardInterrupt: 

With np
1.3626985549926758
1.3909518718719482
1.426476001739502
1.3307390213012695
1.3294603824615479
With PD and size
1.3576700687408447
1.3703715801239014
1.4107935428619385
1.35439133644104
1.3607184886932373
With PD and pd.DataFrame.empty
1.3310844898223877
1.3648407459259033
1.3740041255950928
1.3264544010162354
1.3400328159332275
With np using predefined array size
1.3491873741149902
1.3704020977020264
1.3802015781402588
1.3316471576690674
1.3416483402252197

In [39]:
d_lab_count_arr = []
for lab in unique_labs:
    df_lab = d_lab_file[d_lab_file['labname'] == lab]
    lab_count = df_lab['patientunitstayid'].value_counts()
    stats = get_stats(lab, lab_count.to_numpy())
    d_lab_count_arr.append(stats)
d_count_df = pd.DataFrame(d_lab_count_arr, columns = ["lab", "mean", "var", "median", "IQR", "min", "max", "n"])

In [ ]:
count_df.to_csv(file_path.joinpath('LabCount.csv'))
d_count_df.to_csv(file_path.joinpath('LabCount_delirium.csv'))

In [59]:
# freq_stats = []
rev_freq_stats = []
for lab in unique_labs:
    print(lab)
    df_lab = lab_file[lab_file['labname'] == lab]
    lab_count = df_lab['patientunitstayid'].value_counts()
    # freq_arr = []
    rev_freq_arr = []
    for id, count in lab_count.iteritems():
        df_id = df_lab[df_lab['patientunitstayid'] == id]

        # times = np.sort(df_id['labresultoffset'].to_numpy())
        # time_diff = times[1:] - times[:-1]
        # time_diff = time_diff[time_diff != 0]
        # freq_arr.extend(time_diff)

        rev_times = np.sort(df_id['labresultrevisedoffset'].to_numpy())
        rev_time_diff = rev_times[1:] - rev_times[:-1]
        rev_time_diff = rev_time_diff[rev_time_diff != 0]
        rev_freq_arr.extend(rev_time_diff)

    # freq_stats.append(get_stats(lab, pd.DataFrame(freq_arr)))
    rev_freq_stats.append(get_stats(lab, pd.DataFrame(rev_freq_arr)))

rev_freq_df = pd.DataFrame(rev_freq_stats)
    
        
        


/mnt/Research/LCICM/Delirium/delirium/.venv/lib64/python3.10/site-packages/numpy/core/fromnumeric.py:3440: RuntimeWarning: Mean of empty slice.
  return _methods._mean(a, axis=axis, dtype=dtype,
/mnt/Research/LCICM/Delirium/delirium/.venv/lib64/python3.10/site-packages/numpy/core/_methods.py:189: RuntimeWarning: invalid value encountered in double_scalars
  ret = ret.dtype.type(ret / rcount)


In [22]:
# d_freq_stats = []
d_rev_freq_stats = []
for lab in unique_labs:
    
    df_lab = d_lab_file[d_lab_file['labname'] == lab]
    lab_count = df_lab['patientunitstayid'].value_counts()
    # freq_arr = []
    rev_freq_arr = []
    for id, count in lab_count.iteritems():
        df_id = df_lab[df_lab['patientunitstayid'] == id]

        # times = np.sort(df_id['labresultoffset'].to_numpy())
        # time_diff = times[1:] - times[:-1]
        # time_diff = time_diff[time_diff != 0]
        # freq_arr.extend(time_diff)

        rev_times = np.sort(df_id['labresultrevisedoffset'].to_numpy(dtype=int))
        rev_times = rev_times[rev_times > 0]
        # print(id, type(rev_times))
        #required because squeezing one row into a series turns 
        rev_time_diff = rev_times[1:] - rev_times[:-1]
        rev_time_diff = rev_time_diff[rev_time_diff != 0]
        rev_freq_arr.extend(rev_time_diff)

        # print(rev_times)


    # d_freq_stats.append(get_stats(lab, pd.DataFrame(freq_arr)))
    d_rev_freq_stats.append(get_stats(lab, pd.DataFrame(rev_freq_arr)))
d_rev_freq_df = pd.DataFrame(d_rev_freq_stats, columns=["lab", "mean", "var", "median", "IQR", "min", "max", "n"])

[-47319570      -350      -160        43       281       588       768
      1251      1832      2235      2472      2683      3013      4125
      4593      4887      5565      7348      8484     10210     13028
     13041     14256     18869     19095] [47319220      190      203      238      307      180      483      581
      403      237      211      330     1112      468      294      678
     1783     1136     1726     2818       13     1215     4613      226]
[-47319570      -350      -160        43       281       588       768
      1251      1832      2235      2472      2683      3013      4125
      4593      4887      5565      7348      8484     10210     13028
     13041     14256     18869     19095] [47319220      190      203      238      307      180      483      581
      403      237      211      330     1112      468      294      678
     1783     1136     1726     2818       13     1215     4613      226]
[-47319570      -350      -160        43       281

6.18 min using list appending and np array subtraction

In [23]:
#rev_freq_df = pd.DataFrame(rev_freq_stats, columns=["lab", "mean", "var", "median", "IQR", "min", "max", "n"])
d_rev_freq_df.to_csv(file_path.joinpath('FreqPrevalence_delirium.csv'))
print(d_rev_freq_df)

                      lab                             mean  \
0              fibrinogen  0    1638.925429
dtype: float64   
1                PT - INR  0    1972.715887
dtype: float64   
2               magnesium  0    1636.070943
dtype: float64   
3                      PT  0    1974.185433
dtype: float64   
4                      pH  0    1992.917957
dtype: float64   
..                    ...                              ...   
153       Amikacin - peak       0    2218.5
dtype: float64   
154            Vent Other   0    2121.27957
dtype: float64   
155  HSV 1&2 IgG AB titer                              NaN   
156            HIV 1&2 AB                              NaN   
157        HSV 1&2 IgG AB                              NaN   

                                  var  median      IQR                    min  \
0    0    1.364972e+07
dtype: float64   571.5  1113.25    0    1
dtype: int64   
1    0    1.468596e+07
dtype: float64  1413.0   706.00    0    1
dtype: int64   
2    0    4.